game/src/ecs/systems/HealthSystem.java

```diff
@@ -0,0 +1,64 @@
+import ecs.components.MissingComponentException;
+import ecs.damage.Damage;
+import ecs.entities.Entity;
+import java.util.List;
+import mydungeon.ECS;
+
+/**
+ * The HealthSystem offsets the damage to be done to all entities with the HealthComponent. Triggers
+ * the death of an entity when the hitpoints have fallen below 0.
+ */
+public class HealthSystem extends ECS_System {
+    @Override
+    public void update() {
+        for (Entity entity : ECS.entities) {
+            entity.getComponent(HealthComponent.name)
+                    .ifPresent(
+                            component -> {
+                                {
+                                    final AnimationComponent ac =
+                                            (AnimationComponent)
+                                                    entity.getComponent(AnimationComponent.name)
+                                                            .orElseThrow(
+                                                                    () ->
+                                                                            new MissingComponentException(
+                                                                                    AnimationComponent
+                                                                                            .name));
+
+                                    HealthComponent hpComponent = (HealthComponent) component;
+
+                                    // is the entity dead?
+                                    if (hpComponent.getCurrentHitPoints() <= 0) {
+                                        hpComponent.triggerOnDeath();
+                                        ac.setCurrentAnimation(hpComponent.getDieAnimation());
+                                        // todo: After animation is finished
+                                        ECS.entitiesToRemove.add(entity);
+                                    } else {
+                                        // make damage
+                                        List<Damage> damageToGet = hpComponent.getDamageList();
+                                        int dmgAmmount = 0;
+                                        for (Damage dmg : damageToGet) {
+                                            // place to increase or decrease dmg based on skills,
+                                            // items
+                                            dmgAmmount += dmg.damageAmmount();
+                                        }
+                                        // if damage was caused, play getHitAnimation
+                                        if (dmgAmmount > 0) {
+                                            ac.setCurrentAnimation(
+                                                    hpComponent.getGetHitAnimation());
+                                        }
+                                        // clear list
+                                        damageToGet.clear();
```

Comment:
Was genau soll hier bezweckt werden? Die Liste ist eine lokale Variable - die ist im nächsten Schleifendurchlauf doch eh wieder weg, oder?

Reply:
Wrong. Die ist aus dem component geholt (call by reference)

Reply:
Öhmmm. Ja. Schöner Nebeneffekt :/

Sollte das nicht eher sowas wie `hpComponent.clearDamages()` sein?

Reply:
Öhmmm. Ja. Schöner Nebeneffekt :/

Sollte das nicht eher sowas wie `hpComponent.clearDamages()` sein?

Reply:
Und bei der Gelegenheit könnte es sowas wie `hpComponent.getDamageAmount()` geben?

Reply:
> > Öhmmm. Ja. Schöner Nebeneffekt :/
> > Sollte das nicht eher sowas wie hpComponent.clearDamages() sein?
> 
> Ehrlich gesagt, keine Ahnung was hier der beste style ist. Ich weiß das Spotless jegliches Call by Reference Verhalten doof findet und am liebsten nur kopien von Listen rausgeben will, auf der anderen Seite ist das ganze Konzept von ECS, dass die Systeme die Werte verändern, warum also nicht so? 

Gute Frage. 

Ich sehe hier einen Nebeneffekt - sowas ist immer hässlich. Du siehst ja, ich bin auch beim Lesen drüber gestolpert. 

Das Rausgeben der Liste an sich ist nicht das Problem, aber das Verändern von Werten im Objekt über die herausgegebene Liste ist das Unschöne hier.



> Mit der Argumentation könnte ich aber auch gleich alles public machen und das will nun wirklich keiner.....

Ist tatsächlich aber die Frage, ob das nicht eine konsequente Folge von ECS ist. Im ShPD war auch fast alles `public`, wenn ich mich richtig erinnere.

Und letztlich stellt sich die Frage: Warum willst Du Dinge kapseln und verstecken (`private`), wenn Du Dir dann Getter bastelst und damit exakt die interne Repräsentation rausgibts und dann obendrein noch die Veränderung des Objektzustands darüber erlaubst? Dann kannst Du auch konsequent sein und direkt auf den Attributen des Objekts arbeiten, das macht dann keinen Unterschied mehr.


> > Und bei der Gelegenheit könnte es sowas wie hpComponent.getDamageAmount() geben?
> 
> Nope. Das Component weiß das nämlich nicht. Es kennt ja nicht die Rüstung etc.

Die kenne ich doch hier aber auch nicht? 

Was ich hier im Code sehe: Ich lasse mit vom Component die Liste der Damage-Objekte geben, summiere deren Schadenspunkte auf (unabhängig vom konkreten Typ) und lösche dann in der Component die Liste der Damage-Objekte. 

Warum kann genau das nicht auch die Component machen? Sie hat ja bereits die Liste der Damage-Objekte, und darüber iterieren und was zusammenaddieren kann sie dann auch? Und die Liste wieder löschen? Oder sehe ich hier irgendwas nicht?

Reply:
Ich hab den Code und das Kommentar mal angepasst, wird dir die Idee jetzt klarer?

Reply:
> Und letztlich stellt sich die Frage: Warum willst Du Dinge kapseln und verstecken (private), wenn Du Dir dann Getter bastelst und damit exakt die interne Repräsentation rausgibts und dann obendrein noch die Veränderung des Objektzustands darüber erlaubst? Dann kannst Du auch konsequent sein und direkt auf den Attributen des Objekts arbeiten, das macht dann keinen Unterschied mehr.

Früher hat man (mir) mal gesagt, du benutzt setter und getter (auch), damit im Falle von Umbenennung der Attribute die API sich nicht ändert.

Reply:
Ja, stabile API ist _ein_ Argument. Aber das Hauptargument ist eigentlich die Kapselung - Objekte sollen nur über Methoden zugegriffen und geändert werden. Das geht in "Modern Java" eben dann so weit, dass man Objekt nicht mehr veränderbar hat und bei Änderung einfach ein neues erzeugt.

Reply:
Mit dem Ticket #285 kann hier eigentlich "resolve", oder? Das braucht nochmal Nachdenken und vermutlich Änderungen an mehr als dieser Stelle?